fs/fs.go

```diff
@@ -74,20 +74,68 @@ import (
 	"github.com/opencontainers/go-digest"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+	"golang.org/x/sync/semaphore"
 )
 
 var (
 	defaultIndexSelectionPolicy = SelectFirstPolicy
 	fusermountBin               = "fusermount"
+	layerQueueBufferSize        = 1024
 )
 
+type layerQueue struct {
+	queue chan func() string
```

Comment:
I think the func here should take a context so we pass in the FileSystem context. This way we can cancel in-flight pre-resolves as-well.